row_transform/row_transform.tex

```diff
@@ -0,0 +1,188 @@
+def make_row_by_row(atom_input_domain : DIA, atom_output_domain : DOA, atom_function : F):
+    input_domain = VectorDomain(DIA);
+    output_domain = VectorDomain(DOA)
+    input_metric = SymmetricDistance()
+    output_metric = SymmetricDistance()
+
+    def Relation(d_in : u32, d_out : u32) -> bool: |\label{line:rel}|
+        return d_out <= d_in*1
+    
+    def function(data : Vec[DIA]) -> Vec[DOA]:  |\label{line:fn}|
+        return list(map(atom_function, data)) |\label{line:map}|
+    
+    return Transformation(input_domain, output_domain, function, input_metric, output_metric, stability_relation=Relation)
+
+\end{lstlisting}
+
+
+\section{Proof}
+The necessary definitions for the proof can be found at \href{https://www.overleaf.com/project/60d214e390b337703d200982}{``List of definitions used in the proofs"}.
+
+\begin{theorem}
+    For every setting of the input parameters \texttt{(atom\_input\_domain, atom\_output\_domain, atom\_function)} to \texttt{make\_row\_by\_row} such that the given preconditions
+    hold, the transformation returned by \texttt{make\_row\_by\_row} has the following properties:
+    \begin{enumerate}
+        \item \textup{(Appropriate output domain).} For every element $v$ in \texttt{input\_domain}, $\function(v)$ is in \texttt{output\_domain}. 
+        
+        \item \textup{(Domain-metric compatibility).} The domain \texttt{input\_domain} matches one of the possible domains listed in the definition of \texttt{input\_metric}, and likewise \texttt{output\_domain} matches one of the possible domains listed in the definition of \texttt{output\_metric}.
+        
+        \item \textup{(Stability guarantee).} For every pair of elements $v, w$ in \texttt{input\_domain} and for every pair $(\din, \dout)$, where $\din$ is of the associated type for \texttt{input\_metric} and $\dout$ is the associated type for \texttt{output\_metric}, if $v,w$ are $\din$-close under \texttt{input\_metric} and $\Relation(\din, \dout) = \True$, then $\function(v), \function(w)$ are $\dout$-close under \texttt{output\_metric}.
+    \end{enumerate}
+\end{theorem}
+\begin{proof}
+\begin{enumerate}
+\item \textbf{(Appropriate output domain).} In the case of \texttt{make\_row\_by\_row}, this corresponds to showing that for every vector $v$ of elements of type \texttt{DIA}, \texttt{function}$(v)$ is a vector of elements of type \texttt{DOA}. 
+
+The $\function(v)$ has type \texttt{Vec[DOA]} follows from the assumption that element $v$ is in \texttt{input\_domain} and from the type signature of \texttt{function} in line~\ref{line:fn} of the pseudocode (Section~\ref{sec:pseudocode}), which takes in an element of type \texttt{Vec(DIA)} and returns an element of type \texttt{Vec[DOA]}. If the Rust code compiles correctly, then the type correctness follows from the definition of the type signature enforced by Rust. Otherwise, the code raises a compile time error for incorrect function input type or output type. 
```

Comment:
This doesn't follow solely from the type signature, which only captures the carrier types.  Suppose DIA =DOA=IntervalDomain(u32,L,U).  Then the type signature of f is u32->u32, which does not sure that its outputs are in the interval [L,U] (which might in turn rely on the fact that the inputs are in the interval [L,U]).

Reply:
Good point, will take this into account when rewriting.